File: scripts/impact_brief.py
#!/usr/bin/env python3
"""One-shot impact brief — the cheap path for small/medium changes.

The full map (Layer 1 + Layer 2 JSON) is bigger than a small raw diff, so for
small changes "read the map instead of the diff" loses. But the model can read a
small diff itself for free; what it CAN'T cheaply compute are the cross-repo
signals:

  * blast radius   - who imports each changed public symbol (whole-repo scan).
  * breaking change - did a kept public symbol's signature change (base vs head
                      AST compare).
  * test gap       - code changed with no test files touched.

This orchestrates the existing tools and distills *only* those signals into a
few hundred bytes — net-positive even on a one-file change, because it adds
information the model would otherwise have to grep the repo / diff two revisions
to get, at negligible context cost. Read the small diff yourself; use this for
the impact.

Usage:
    python scripts/impact_brief.py --range main..HEAD --cwd <repo>
    python scripts/impact_brief.py --staged --cwd <repo>      # no breaking info
    python scripts/impact_brief.py --range main..HEAD --cwd <repo> --json

Exit 0 always (degrades gracefully: missing grammars just drop blast/breaking).
"""

import argparse
import json
import os
import subprocess
import sys
import tempfile

SCRIPTS = os.path.dirname(os.path.abspath(__file__))
TOP_IMPACT = 5


def run_json(script, *args):
    """Run a sibling script; return parsed JSON (or None on non-JSON/empty)."""
    cp = subprocess.run([sys.executable, os.path.join(SCRIPTS, script), *map(str, args)],
                        stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
    out = cp.stdout.strip()
    try:
        return json.loads(out)
    except ValueError:
        return None


def main():
    p = argparse.ArgumentParser(description="One-shot cross-repo impact brief")
    p.add_argument("--range", help="git range base..head")
    p.add_argument("--staged", action="store_true", help="brief the staged index")
    p.add_argument("--cwd", default=".", help="repository working directory")
    p.add_argument("--json", action="store_true", help="emit JSON instead of text")
    args = p.parse_args()

    cwd = os.path.abspath(args.cwd)

    # Layer 1 (always works: Python + git).
    l1_args = ["diff_summary.py", "--cwd", cwd, "--compact"]
    if args.staged:
        l1_args.append("--staged")
    elif args.range:
        l1_args += ["--range", args.range]
    l1 = run_json(*l1_args)
    if l1 is None:
        sys.stderr.write("error: could not read the diff\n")
        sys.exit(2)

    totals = l1.get("totals", {})
    flags = [f for f in l1.get("risk_flags", [])
             if f in ("code_changed_without_tests", "contains_deletions")]

    # Layer 2 blast radius (needs the L1 file list; may fall back).
    high_impact = []
    review_order = l1.get("review_order", [])
    with tempfile.NamedTemporaryFile("w", suffix=".json", delete=False) as fh:
        json.dump(l1, fh)
        l1_path = fh.name
    try:
        l2 = run_json("symbol_impact.py", "--root", cwd, "--diff-json", l1_path, "--compact")
    finally:
        os.unlink(l1_path)
    if l2 and l2.get("ok"):
        review_order = l2.get("review_order", review_order)
        public = [s for s in l2.get("symbols", []) if s.get("exported")]
        public.sort(key=lambda s: -s.get("blast_radius", 0))
        for s in public[:TOP_IMPACT]:
            if s.get("blast_radius", 0) > 0:
                high_impact.append({"symbol": s["name"], "file": s["file"],
                                    "blast_radius": s["blast_radius"]})

    # Breaking changes need two committed revisions (refactor_check).
    breaking = []
    if args.range:
        rc = run_json("refactor_check.py", "--range", args.range, "--cwd", cwd)
        if rc and rc.get("ok"):
            inv = rc.get("invariants", {})
            breaking = sorted(set(inv.get("public_signatures_changed", []))
                              | set(inv.get("public_api_removed", [])))

    brief = {
        "files": totals.get("files", 0),
        "additions": totals.get("additions", 0),
        "deletions": totals.get("deletions", 0),
        "flags": flags,
        "breaking_changes": breaking,
        "high_impact": high_impact,
        "review_order": review_order[:TOP_IMPACT],
    }

    if args.json:
        json.dump(brief, sys.stdout, indent=2)
        sys.stdout.write("\n")
        return

    # Compact human line(s).
    head = "%d file(s), +%d/-%d" % (brief["files"], brief["additions"], brief["deletions"])
    bits = [head]
    if "code_changed_without_tests" in flags:
        bits.append("no tests touched")
    if breaking:
        bits.append("BREAKING: " + ", ".join(breaking))
    print(" · ".join(bits))
    if high_impact:
        print("high impact:")
        for h in high_impact:
            print("  %s (%s) — %d caller file(s)" % (h["symbol"], h["file"], h["blast_radius"]))
    if not breaking and not high_impact:
        print("no public-API impact or breaking changes detected")


if __name__ == "__main__":
    main()

File: js/impact_brief.mjs
#!/usr/bin/env node
// One-shot impact brief (small-change path) — Node port of scripts/impact_brief.py.
// Orchestrates the sibling scripts and distills only the cross-repo signals.

import { execFileSync } from "node:child_process";
import { writeFileSync, unlinkSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { pathToFileURL } from "node:url";

const JS = path.dirname(new URL(import.meta.url).pathname);
const TOP_IMPACT = 5;

function runJson(script, args) {
  let out;
  try { out = execFileSync("node", [path.join(JS, script), ...args], { encoding: "utf8", maxBuffer: 1 << 28 }); }
  catch (e) { out = e.stdout || ""; }
  try { return JSON.parse(out.trim()); } catch { return null; }
}
function runText(script, args) {
  try { return execFileSync("node", [path.join(JS, script), ...args], { encoding: "utf8", maxBuffer: 1 << 28 }); }
  catch (e) { return e.stdout || ""; }
}

function parseArgs(argv) {
  const a = { range: null, staged: false, cwd: ".", json: false };
  for (let i = 0; i < argv.length; i++) {
    const x = argv[i];
    if (x === "--range") a.range = argv[++i];
    else if (x === "--staged") a.staged = true;
    else if (x === "--cwd") a.cwd = argv[++i];
    else if (x === "--json") a.json = true;
  }
  return a;
}

function main() {
  const a = parseArgs(process.argv.slice(2));
  const cwd = path.resolve(a.cwd);
  const scope = a.staged ? ["--staged"] : (a.range ? ["--range", a.range] : []);

  const l1 = runJson("diff_summary.mjs", ["--cwd", cwd, "--compact", ...scope]);
  if (!l1) { process.stderr.write("error: could not read the diff\n"); process.exit(2); }

  const totals = l1.totals || {};
  const flags = (l1.risk_flags || []).filter((f) => f === "code_changed_without_tests" || f === "contains_deletions");

  let highImpact = [];
  let reviewOrder = l1.review_order || [];
  const l1Path = path.join(tmpdir(), `crd-l1-${process.pid}-${Date.now()}.json`);
  writeFileSync(l1Path, JSON.stringify(l1));
  let l2;
  try { l2 = runJson("symbol_impact.mjs", ["--root", cwd, "--diff-json", l1Path, "--compact"]); }
  finally { try { unlinkSync(l1Path); } catch { /* */ } }
  if (l2 && l2.ok) {
    reviewOrder = l2.review_order || reviewOrder;
    const pub = (l2.symbols || []).filter((s) => s.exported).sort((x, y) => y.blast_radius - x.blast_radius);
    for (const s of pub.slice(0, TOP_IMPACT)) {
      if (s.blast_radius > 0) highImpact.push({ symbol: s.name, file: s.file, blast_radius: s.blast_radius });
    }
  }

  let breaking = [];
  if (a.range) {
    const rc = runJson("refactor_check.mjs", ["--range", a.range, "--cwd", cwd]);
    if (rc && rc.ok) {
      const inv = rc.invariants || {};
      breaking = [...new Set([...(inv.public_signatures_changed || []), ...(inv.public_api_removed || [])])].sort();
    }
  }

  const brief = {
    files: totals.files || 0,
    additions: totals.additions || 0,
    deletions: totals.deletions || 0,
    flags, breaking_changes: breaking, high_impact: highImpact,
    review_order: reviewOrder.slice(0, TOP_IMPACT),
  };

  if (a.json) { process.stdout.write(JSON.stringify(brief, null, 2) + "\n"); return; }

  const bits = [`${brief.files} file(s), +${brief.additions}/-${brief.deletions}`];
  if (flags.includes("code_changed_without_tests")) bits.push("no tests touched");
  if (breaking.length) bits.push("BREAKING: " + breaking.join(", "));
  console.log(bits.join(" · "));
  if (highImpact.length) {
    console.log("high impact:");
    for (const h of highImpact) console.log(`  ${h.symbol} (${h.file}) — ${h.blast_radius} caller file(s)`);
  }
  if (!breaking.length && !highImpact.length) console.log("no public-API impact or breaking changes detected");
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) main();
